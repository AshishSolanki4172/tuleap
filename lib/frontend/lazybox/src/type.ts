/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { UpdateFunction } from "hybrids";
import type { HTMLTemplateStringProcessor } from "./index";
import type { SearchInput } from "./SearchInput";
import type { SelectionBadge } from "./selection/SelectionBadge";
import type { SelectionElement } from "./selection/SelectionElement";
import type { DropdownElement } from "./dropdown/DropdownElement";

export interface Lazybox {
    setDropdownContent(groups: GroupCollection): void;
    resetSelection(): void;
    replaceSelection(selection: ReadonlyArray<LazyboxItem>): void;
    destroy(): void;
}

export type LazyboxSearchInputCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type HTMLTemplateResult = UpdateFunction<HTMLElement>;
export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyboxSelectionBadgeCallback = (item: LazyboxItem) => SelectionBadge & HTMLElement;

type LazyboxWithMultipleSelection = {
    readonly is_multiple: true;
    readonly selection_badge_callback?: LazyboxSelectionBadgeCallback;
};

type LazyboxWithSingleSelection = {
    readonly is_multiple: false;
    readonly search_input_placeholder: string;
};

type LazyboxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyboxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyboxWithSingleSelection | LazyboxWithMultipleSelection) &
    (LazyboxWithoutNewItemButton | LazyboxWithNewItemButton) & {
        readonly placeholder: string;
        readonly templating_callback: LazyboxTemplatingCallback;
        readonly selection_callback: LazyboxSelectionCallback;
        readonly search_input_callback: LazyboxSearchInputCallback;
    };

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement & DropdownElement;
    search_field_element: HTMLElement & SearchInput;
    selection_element: HTMLElement & SelectionElement;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
