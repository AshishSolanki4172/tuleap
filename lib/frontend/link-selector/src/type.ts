/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
