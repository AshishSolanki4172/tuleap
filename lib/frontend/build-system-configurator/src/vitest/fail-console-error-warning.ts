/**
 * Copyright (c) Enalean, 2022-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { beforeEach, afterEach, vi } from "vitest";

let is_console_error_or_warning = false;

beforeEach(() => {
    is_console_error_or_warning = false;
    const originalError = global.console.error;
    vi.spyOn(global.console, "error").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalError(...args);
    });
    const originalWarn = global.console.warn;
    vi.spyOn(global.console, "warn").mockImplementation((...args: unknown[]) => {
        is_console_error_or_warning = true;
        originalWarn(...args);
    });
});

afterEach(() => {
    if (is_console_error_or_warning) {
        throw new Error("Console warnings and errors are not allowed");
    }
});
