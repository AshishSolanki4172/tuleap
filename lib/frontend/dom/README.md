# DOM

Common utility functions to manipulate the DOM in Vanilla TypeScript.
