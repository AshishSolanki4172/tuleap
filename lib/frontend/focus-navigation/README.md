# focus-navigation

Functions to help with keyboard navigation between related elements. Move the focus up or down in a common set of elements. To group elements in a common set, assign them a `data-navigation=<name of the set>` attribute. For example: `data-navigation=card`

_Note:_ This library _does not_ register keyboard event handlers, that's up to the caller.
