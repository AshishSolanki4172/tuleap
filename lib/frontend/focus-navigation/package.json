{
  "author": "Enalean Team",
  "name": "@tuleap/focus-navigation",
  "homepage": "https://tuleap.org",
  "license": "GPL-2.0-or-later",
  "private": true,
  "type": "module",
  "module": "dist/focus-navigation.js",
  "main": "dist/focus-navigation.umd.cjs",
  "types": "dist/main.d.ts",
  "exports": {
    ".": {
      "import": "./dist/focus-navigation.js",
      "require": "./dist/focus-navigation.umd.cjs"
    }
  },
  "devDependencies": {
    "@tuleap/build-system-configurator": "workspace:*",
    "@tuleap/dom": "workspace:*"
  },
  "scripts": {
    "typecheck": "tsc --noEmit",
    "build": "vite build",
    "watch": "vite build --watch --mode development --minify false",
    "test": "vitest"
  }
}
