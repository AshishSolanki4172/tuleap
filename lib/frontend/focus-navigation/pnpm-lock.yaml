lockfileVersion: 5.4

specifiers:
  '@tuleap/build-system-configurator': workspace:*
  '@tuleap/dom': workspace:*

devDependencies:
  '@tuleap/build-system-configurator': link:../build-system-configurator
  '@tuleap/dom': link:../dom
