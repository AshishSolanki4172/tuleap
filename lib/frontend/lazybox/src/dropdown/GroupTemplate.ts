/*
 * Copyright (c) Enalean, 2023-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { UpdateFunction } from "hybrids";
import { html } from "hybrids";
import type { DropdownElement } from "./DropdownElement";
import type { GroupOfItems, LazyboxItem } from "../GroupCollection";
import { isEnterKey } from "../helpers/keys-helper";
import { onArrowKeyDown, onArrowKeyUp } from "./DropdownElement";

export const getItemTemplate = (
    host: DropdownElement,
    item: LazyboxItem
): UpdateFunction<DropdownElement> => {
    if (item.is_disabled) {
        return html`<li
            role="option"
            class="lazybox-dropdown-option-value-disabled"
            data-test="lazybox-item"
        >
            ${host.templating_callback(html, item)}
        </li>`;
    }

    const onPointerUp = (host: DropdownElement): void => {
        host.selection.selectItem(item);
        host.open = false;
    };

    const onKeyUp = (host: DropdownElement & HTMLElement, event: KeyboardEvent): void => {
        if (isEnterKey(event)) {
            event.stopPropagation();
            host.selection.selectItem(item);
            host.open = false;
            return;
        }
        onArrowKeyUp(host, event);
    };

    const aria_value = String(host.selection.isSelected(item));
    return html`<li
        role="option"
        tabindex="0"
        aria-selected="${aria_value}"
        class="lazybox-dropdown-option-value"
        data-test="lazybox-item"
        data-navigation="lazybox-item"
        onpointerup="${onPointerUp}"
        onkeyup="${onKeyUp}"
        onkeydown="${onArrowKeyDown}"
    >
        ${host.templating_callback(html, item)}
    </li>`;
};

const createEmptyDropdownState = (dropdown_message: string): UpdateFunction<DropdownElement> =>
    html`<li
        class="lazybox-empty-dropdown-state"
        role="alert"
        aria-live="assertive"
        data-test="lazybox-empty-state"
    >
        ${dropdown_message}
    </li>`;

const getGroupLabel = (group: GroupOfItems): UpdateFunction<DropdownElement> => {
    if (group.is_loading) {
        return html`<strong class="lazybox-group-label">
            ${group.label}
            <i
                class="fa-solid fa-spin fa-circle-notch lazybox-loading-group-spinner"
                data-test="lazybox-loading-group-spinner"
            ></i>
        </strong>`;
    }

    return html`<strong class="lazybox-group-label">${group.label}</strong>`;
};

const getGroupFooter = (group: GroupOfItems): UpdateFunction<DropdownElement> =>
    group.footer_message !== "" && group.items.length > 0
        ? html`<p class="lazybox-group-footer" data-test="lazybox-group-footer">
              ${group.footer_message}
          </p>`
        : html``;

const getGroupTemplate = (
    host: DropdownElement,
    group: GroupOfItems
): UpdateFunction<DropdownElement> => {
    const item_templates =
        group.items.length === 0
            ? [createEmptyDropdownState(group.empty_message)]
            : group.items.map((item) => getItemTemplate(host, item));

    return html`<li class="lazybox-item-group">
        ${getGroupLabel(group)}
        <ul role="group" aria-label="${group.label}" class="lazybox-dropdown-option-group">
            ${item_templates}
        </ul>
        ${getGroupFooter(group)}
    </li>`;
};

export const getAllGroupsTemplate = (host: DropdownElement): UpdateFunction<DropdownElement> =>
    html`${host.groups.map((group) => getGroupTemplate(host, group))}`;
