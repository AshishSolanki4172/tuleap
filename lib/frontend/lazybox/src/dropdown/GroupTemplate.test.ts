/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { beforeEach, describe, expect, it, vi } from "vitest";
import { GroupCollectionBuilder } from "../../tests/builders/GroupCollectionBuilder";
import type { GroupCollection, LazyboxItem } from "../GroupCollection";
import { TemplatingCallbackStub } from "../../tests/stubs/TemplatingCallbackStub";
import { getAllGroupsTemplate, getItemTemplate } from "./GroupTemplate";
import type { HostElement } from "./DropdownElement";
import { LazyboxItemStub } from "../../tests/stubs/LazyboxItemStub";
import { selectOrThrow } from "@tuleap/dom";
import * as tuleap_focus from "@tuleap/focus-navigation";

vi.mock("@tuleap/focus-navigation");

const noopSelectItem = (item: LazyboxItem): void => {
    if (item) {
        //Do nothing
    }
};

describe("GroupTemplate", () => {
    let target: ShadowRoot, groups: GroupCollection;

    beforeEach(() => {
        const doc = document.implementation.createHTMLDocument();
        target = doc.createElement("div") as unknown as ShadowRoot;

        groups = [];
    });

    const getHost = (): HostElement =>
        ({
            open: true,
            groups,
            selection: {
                selectItem: noopSelectItem,
                isSelected: (item) => (item ? false : false),
            },
            templating_callback: TemplatingCallbackStub.build(),
        }) as HostElement;

    const render = (): void => {
        const host = getHost();
        const updateFunction = getAllGroupsTemplate(host);
        updateFunction(host, target);
    };

    describe("rendering", () => {
        it("renders grouped list items", () => {
            groups = GroupCollectionBuilder.withTwoGroups();
            render();
            expect(target.innerHTML).toMatchSnapshot();
        });

        it("renders empty option groups when they have a placeholder", () => {
            const empty_state_text = "No results found on the server";
            groups = GroupCollectionBuilder.withSingleGroup({
                empty_message: empty_state_text,
                items: [],
            });

            render();

            const empty_state = selectOrThrow(target, "[data-test=lazybox-empty-state]");
            expect(empty_state.textContent?.trim()).toBe(empty_state_text);
        });

        it(`renders group footer message below the list of items`, () => {
            const footer_message = "Maybe there are more results";
            groups = GroupCollectionBuilder.withSingleGroup({
                items: [
                    LazyboxItemStub.withDefaults({ value: { id: 1 } }),
                    LazyboxItemStub.withDefaults({ value: { id: 2 } }),
                ],
                footer_message,
            });

            render();

            const footer = selectOrThrow(target, "[data-test=lazybox-group-footer]");
            expect(footer.textContent?.trim()).toBe(footer_message);
        });

        it("renders a spinner next to the group title when it is loading", () => {
            const empty_state_text = "I am loading, wait a second!";
            groups = GroupCollectionBuilder.withSingleGroup({
                label: "A group still loading",
                empty_message: empty_state_text,
                items: [],
                is_loading: true,
            });

            render();

            const spinner = target.querySelector("[data-test=lazybox-loading-group-spinner]");
            expect(spinner).toBeDefined();
        });
    });

    describe(`getItemTemplate()`, () => {
        let item: LazyboxItem;

        beforeEach(() => {
            item = LazyboxItemStub.withDefaults({ value: { id: 1 } });
        });

        const render = (host: HostElement): void => {
            const updateFunction = getItemTemplate(host, item);
            updateFunction(host, target);
        };

        it(`renders a list item from a given template`, () => {
            render(getHost());
            expect(target.innerHTML).toMatchInlineSnapshot(`
              <li role="option" tabindex="0" class="lazybox-dropdown-option-value" data-navigation="lazybox-item" aria-selected="false">
                <span>Badge</span>Value 1
              </li>
            `);
        });

        it(`when I click on a list item,
            it will select the item and close the dropdown`, () => {
            const host = getHost();
            const selectItem = vi.spyOn(host.selection, "selectItem");

            render(host);
            const list_item = selectOrThrow(target, "[data-test=lazybox-item]");
            list_item.dispatchEvent(new Event("pointerup"));

            expect(selectItem).toHaveBeenCalled();
            expect(host.open).toBe(false);
        });

        it(`when I press the "enter" key while focusing a list item,
            it will select the item and close the dropdown
            and will stop propagation (it has already been handled)`, () => {
            const host = getHost();
            const selectItem = vi.spyOn(host.selection, "selectItem");

            render(host);
            const list_item = selectOrThrow(target, "[data-test=lazybox-item]");
            const event = new KeyboardEvent("keyup", { key: "Enter" });
            const stopPropagation = vi.spyOn(event, "stopPropagation");
            list_item.dispatchEvent(event);

            expect(selectItem).toHaveBeenCalled();
            expect(host.open).toBe(false);
            expect(stopPropagation).toHaveBeenCalled();
        });

        it(`when I press the "arrow down" key while focusing a list item,
            it will prevent default (it will not scroll down)
            and it will focus the next item in the dropdown`, () => {
            const moveFocus = vi.spyOn(tuleap_focus, "moveFocus");
            const key_down_event = new KeyboardEvent("keydown", {
                key: "ArrowDown",
                cancelable: true,
            });
            render(getHost());

            const list_item = selectOrThrow(target, "[data-test=lazybox-item]");
            list_item.dispatchEvent(key_down_event);
            expect(key_down_event.defaultPrevented).toBe(true);

            list_item.dispatchEvent(new KeyboardEvent("keyup", { key: "ArrowDown" }));
            expect(moveFocus.mock.calls[0][1]).toBe("down");
        });

        it(`when I press the "arrow up" key while focusing a list item,
            it will prevent default (it will not scroll up)
            and it will focus the previous item in the dropdown`, () => {
            const moveFocus = vi.spyOn(tuleap_focus, "moveFocus");
            const key_down_event = new KeyboardEvent("keydown", {
                key: "ArrowUp",
                cancelable: true,
            });
            render(getHost());

            const list_item = selectOrThrow(target, "[data-test=lazybox-item]");
            list_item.dispatchEvent(key_down_event);
            expect(key_down_event.defaultPrevented).toBe(true);

            list_item.dispatchEvent(new KeyboardEvent("keyup", { key: "ArrowUp" }));
            expect(moveFocus.mock.calls[0][1]).toBe("up");
        });

        it(`renders a disabled list item`, () => {
            item = LazyboxItemStub.withDefaults({ is_disabled: true });
            render(getHost());

            const item_template = target.firstElementChild;
            expect(
                item_template?.classList.contains("lazybox-dropdown-option-value-disabled"),
            ).toBe(true);
        });
    });
});
