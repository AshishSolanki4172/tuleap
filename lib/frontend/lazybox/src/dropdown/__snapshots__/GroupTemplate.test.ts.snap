// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GroupTemplate > rendering > renders grouped list items 1`] = `
" <li class=\\"lazybox-item-group\\">
        <strong class=\\"lazybox-group-label\\">Group 1</strong>
        <ul role=\\"group\\" class=\\"lazybox-dropdown-option-group\\" aria-label=\\"Group 1\\">
             <li role=\\"option\\" tabindex=\\"0\\" class=\\"lazybox-dropdown-option-value\\" data-test=\\"lazybox-item\\" aria-selected=\\"false\\">
        <span>Badge</span>Value 0
    </li> <li role=\\"option\\" tabindex=\\"0\\" class=\\"lazybox-dropdown-option-value\\" data-test=\\"lazybox-item\\" aria-selected=\\"false\\">
        <span>Badge</span>Value 1
    </li> <li role=\\"option\\" tabindex=\\"0\\" class=\\"lazybox-dropdown-option-value\\" data-test=\\"lazybox-item\\" aria-selected=\\"false\\">
        <span>Badge</span>Value 2
    </li>
        </ul>
        
    </li> <li class=\\"lazybox-item-group\\">
        <strong class=\\"lazybox-group-label\\">Group 2</strong>
        <ul role=\\"group\\" class=\\"lazybox-dropdown-option-group\\" aria-label=\\"Group 2\\">
             <li role=\\"option\\" tabindex=\\"0\\" class=\\"lazybox-dropdown-option-value\\" data-test=\\"lazybox-item\\" aria-selected=\\"false\\">
        <span>Badge</span>Value 3
    </li> <li role=\\"option\\" tabindex=\\"0\\" class=\\"lazybox-dropdown-option-value\\" data-test=\\"lazybox-item\\" aria-selected=\\"false\\">
        <span>Badge</span>Value 4
    </li> <li role=\\"option\\" class=\\"lazybox-dropdown-option-value-disabled\\" data-test=\\"lazybox-item\\">
            <span>Badge</span>Value 5
        </li>
        </ul>
        
    </li>"
`;
